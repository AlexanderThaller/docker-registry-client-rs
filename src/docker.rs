use std::{
    collections::HashMap,
    sync::Arc,
};

use reqwest::{
    header::HeaderMap,
    Client as HTTPClient,
};
use tokio::sync::RwLock;
use url::Url;

use crate::{
    image_name::{
        ImageName,
        Registry,
    },
    manifest::Manifest,
};

mod error;

pub use error::Error;

#[derive(Debug, Default, Clone)]
pub struct Client {
    client: HTTPClient,

    token_cache: Arc<RwLock<HashMap<CacheKey, String>>>,
}

#[derive(Debug, Eq, PartialEq, Hash)]
struct CacheKey {
    registry: Registry,
    repository: String,
    image_name: String,
}

#[derive(Debug, Clone)]
pub struct Response {
    pub digest: Option<String>,
    pub manifest: Manifest,
}

impl Client {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// # Errors
    /// Returns an error if the request fails.
    /// Returns an error if the response body is not valid JSON.
    /// Returns an error if the response body is not a valid manifest.
    /// Returns an error if the response status is not successful.
    pub async fn get_manifest(&self, image_name: &ImageName) -> Result<Response, Error> {
        let mut headers = self.get_headers(image_name).await?;

        let accept_header = [
            "application/vnd.docker.container.image.v1+json",
            "application/vnd.docker.distribution.manifest.list.v2+json",
            "application/vnd.docker.distribution.manifest.v2+json",
            "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
            "application/vnd.docker.plugin.v1+json",
            "application/vnd.oci.image.index.v1+json",
            "application/vnd.oci.image.manifest.v1+json",
        ]
        .join(", ");

        headers.insert(
            "Accept",
            accept_header
                .parse()
                .map_err(Error::ParseManifestAcceptHeader)?,
        );

        let registry_domain = image_name.registry.registry_domain();

        let url = Url::parse(&format!(
            "https://{}/v2/{}/{}/manifests/{}",
            registry_domain, image_name.repository, image_name.image_name, image_name.identifier
        ))
        .map_err(Error::InvalidManifestUrl)?;

        let response = self
            .client
            .get(url.as_str())
            .headers(headers)
            .send()
            .await
            .map_err(Error::GetManifest)?;

        let status = response.status();
        let digest = response
            .headers()
            .get("Docker-Content-Digest")
            .map(|header| {
                header
                    .to_str()
                    .map(String::from)
                    .map_err(Error::ParseDockerContentDigestHeader)
            })
            .transpose()?;

        let body = response.text().await.map_err(Error::ExtractManifestBody)?;

        if !status.is_success() {
            if status == reqwest::StatusCode::NOT_FOUND {
                return Err(Error::ManifestNotFound(url));
            }

            return Err(Error::FailedManifestRequest(status, body));
        }

        let manifest =
            serde_json::from_str(&body).map_err(|e| Error::DeserializeManifestBody(e, body))?;

        Ok(Response { digest, manifest })
    }

    async fn get_headers(&self, image_name: &ImageName) -> Result<HeaderMap, Error> {
        #[derive(Debug, serde::Deserialize)]
        struct Token {
            token: String,
        }

        let cache_key = CacheKey {
            registry: image_name.registry.clone(),
            repository: image_name.repository.clone(),
            image_name: image_name.image_name.clone(),
        };

        let token_cache = self.token_cache.read().await;
        let token = token_cache.get(&cache_key).cloned();
        drop(token_cache);

        let token = if let Some(token) = token {
            token
        } else {
            let token_url = match image_name.registry {
                Registry::Github => format!(
                    "https://ghcr.io/token?scope=repository:{}/{}:pull&service=ghcr.io",
                    image_name.repository, image_name.image_name
                ),
                Registry::DockerHub => format!("https://auth.docker.io/token?service=registry.docker.io&scope=repository:{}/{}:pull&service=registry.docker.io", image_name.repository, image_name.image_name),
                Registry::Quay => format!("https://quay.io/v2/auth?scope=repository:{}/{}:pull&service=quay.io", image_name.repository, image_name.image_name),

                Registry::Specific(_) => return Ok(HeaderMap::new()),
            };

            let token_url = Url::parse(&token_url).map_err(Error::InvalidTokenUrl)?;

            let response = self
                .client
                .get(token_url)
                .send()
                .await
                .map_err(Error::GetToken)?;

            let body = response.text().await.map_err(Error::ExtractTokenBody)?;

            let token: Token =
                serde_json::from_str(&body).map_err(|e| Error::DeserializeToken(e, body))?;

            let mut token_cache = self.token_cache.write().await;
            token_cache.insert(cache_key, token.token.clone());

            token.token
        };

        let mut headers = HeaderMap::new();
        headers.insert(
            "Authorization",
            format!("Bearer {token}")
                .parse()
                .map_err(Error::ParseAuthorizationHeader)?,
        );

        Ok(headers)
    }
}
