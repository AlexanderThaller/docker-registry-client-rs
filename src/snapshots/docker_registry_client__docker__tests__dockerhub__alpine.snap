---
source: src/docker.rs
expression: response
---
{
  "digest": "sha256:765942a4039992336de8dd5db680586e1a206607dd06170ff0a37267a9e01958",
  "manifest": {
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.index.v1+json",
    "manifests": [
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1022,
        "digest": "sha256:008827ed2172a676b08121e21cf9db0ce08a90ee6c8a12fc374af8a56c0e496d",
        "platform": {
          "architecture": "amd64",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:eaa1a55517a9b7caee112a0f5b2fd541f222413234a035b5b89c812efa82137c",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1023,
        "digest": "sha256:af1169cde24ef30cc22ef332bea3d3454302d23bd508184f8761808048070031",
        "platform": {
          "architecture": "arm",
          "os": "linux",
          "variant": "v6"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 566,
        "digest": "sha256:c590d048e7be632c6559d39c16735701bb3bba636a82f32aac6834a349b22ca3",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1023,
        "digest": "sha256:7f94a9bfb82277581c655bacc17bcb07a62b85c6169a1d6ad5df0bde15040e02",
        "platform": {
          "architecture": "arm",
          "os": "linux",
          "variant": "v7"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:b5f0ca7ef2f184953bc22801b7f54293246e220c126f2921313a9b7ea34a5772",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1025,
        "digest": "sha256:ca485c6a7b51db176b2b922aae711edb0cc82953424663297f279babaa3c2fb7",
        "platform": {
          "architecture": "arm64",
          "os": "linux",
          "variant": "v8"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:f1a7e0415ff82b0ad0f3aae3dbc9edebb09e63d332698a953ac0c614bb7076ff",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1018,
        "digest": "sha256:aecf0fcf2640d401102308301e4dcc2fa0acb1608e098ae8e7bc97843cb2ca2f",
        "platform": {
          "architecture": "386",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:77b65d7b3a203c6a06df1b937e7da4fb707b40bd8f424dc892f729182af95b38",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1025,
        "digest": "sha256:35b046b830a6f76d7a06aa2506ae082ab549ff3e720c52274d5e1a561de32f1b",
        "platform": {
          "architecture": "ppc64le",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:bce2e39bac504ae0875b5333680e9d3d0b389f1858fc85b1620161a0e1987eb9",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1025,
        "digest": "sha256:232135d832a25a3b20c0f79e642e2e2d635f8282ee022ca0aa81dc701a97b9b7",
        "platform": {
          "architecture": "riscv64",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:4ebfd85eb89796106d40bff703837b29fe275de7aea0a8e7f05ad77c740e7fad",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 1021,
        "digest": "sha256:38381f79c1fae0f1d58ae9910bb0793c3ad41d3e9496b6de8b9f3e4332d7a3d8",
        "platform": {
          "architecture": "s390x",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 838,
        "digest": "sha256:20a1f9e3f0411b3fa0e3766c53ab26a52c5e15ec2d172fe6058d6f96008f0d20",
        "platform": {
          "architecture": "unknown",
          "os": "unknown"
        }
      }
    ]
  }
}
